enhancements/dns/forwarding.md

```diff
@@ -0,0 +1,246 @@
+---
+title: configurable-dns-forwarding
+authors:
+  - "@dhansen"
+reviewers:
+  - "@ironcladlou"
+  - "@Miciah"
+  - "@frobware"
+approvers:
+  - "@knobunc"
+  - "@ironcladlou"
+creation-date: 2019-09-27
+last-updated: 2019-09-27
+status: implementable
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Configurable DNS Forwarding
+
+This proposal allows cluster operators the ability to configure DNS name query forwarding.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+1. Is `operator.openshift.io` the best API group for the proposed `Upstream` type?
+2. How are API changes handled for cluster downgrades?
+3. Should IPv6 references for `Nameservers` be removed until OpenShift adds IPv6 support?
+
+## Summary
+
+Currently, the [cluster-dns-operator](https://github.com/openshift/cluster-dns-operator) manages DNS forwarding with a
+static ConfigMap:
+
+```bash
+$ oc get configmap/dns-default -n openshift-dns -o yaml
+apiVersion: v1
+data:
+  Corefile: |
+    .:5353 {
+        errors
+        health
+        kubernetes cluster.local in-addr.arpa ip6.arpa {
+            pods insecure
+            upstream
+            fallthrough in-addr.arpa ip6.arpa
+        }
+        prometheus :9153
+        forward . /etc/resolv.conf
+        cache 30
+        reload
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    dns.operator.openshift.io/owning-dns: default
+  name: dns-default
+  namespace: openshift-dns
+  <SNIP>
+```
+
+This proposal will generate this ConfigMap with a `forward` section based on user-provided values of type `Upstream`.
```

Comment:
Note that the generated configuration will need to have a _server block_ with a `forward` section for each upstream, or else you're gonna have a bad time (see https://coredns.io/manual/toc/#forwarding-domains-to-different-upstreams).

Reply:
@Miciah thanks for the heads-up. My next commit will include an example `Upstream` and the associated `ConfigMap` that is generated by the cluster-dns operator.